src/core/tls.c

```diff
@@ -0,0 +1,200 @@
+		tls_rec->certs = NULL;
+	}
+
+	g_free(tls_rec);
+}
+
+void tls_rec_set_protocol_version(TLS_REC *tls_rec, const char *protocol_version)
+{
+	g_assert(tls_rec != NULL);
+
+	tls_rec->protocol_version = g_strdup(protocol_version);
+}
+
+void tls_rec_set_cipher(TLS_REC *tls_rec, const char *cipher)
+{
+	g_assert(tls_rec != NULL);
+
+	tls_rec->cipher = g_strdup(cipher);
+}
+
+void tls_rec_set_cipher_size(TLS_REC *tls_rec, size_t size)
+{
+	g_assert(tls_rec != NULL);
+
+	tls_rec->cipher_size = size;
+}
+
+void tls_rec_set_public_key_algorithm(TLS_REC *tls_rec, const char *algorithm)
+{
+	g_assert(tls_rec != NULL);
+
+	tls_rec->public_key_algorithm = g_strdup(algorithm);
+}
+
+void tls_rec_set_public_key_fingerprint(TLS_REC *tls_rec, const char *fingerprint)
+{
+	g_assert(tls_rec != NULL);
+
+	tls_rec->public_key_fingerprint = g_strdup(fingerprint);
+}
+
+void tls_rec_set_public_key_fingerprint_algorithm(TLS_REC *tls_rec, const char *algorithm)
+{
+	g_assert(tls_rec != NULL);
+
+	tls_rec->public_key_fingerprint_algorithm = g_strdup(algorithm);
+}
+
+void tls_rec_set_public_key_size(TLS_REC *tls_rec, size_t size)
+{
+	g_assert(tls_rec != NULL);
+	tls_rec->public_key_size = size;
+}
+
+void tls_rec_set_not_after(TLS_REC *tls_rec, const char *not_after)
+{
+	g_assert(tls_rec != NULL);
+	tls_rec->not_after = g_strdup(not_after);
+}
+
+void tls_rec_set_not_before(TLS_REC *tls_rec, const char *not_before)
+{
+	g_assert(tls_rec != NULL);
+	tls_rec->not_before = g_strdup(not_before);
+}
+
+#ifdef SSL_get_server_tmp_key
+void tls_rec_set_ephemeral_key_algorithm(TLS_REC *tls_rec, const char *algorithm)
+{
+	g_assert(tls_rec != NULL);
+	tls_rec->ephemeral_key_algorithm = g_strdup(algorithm);
+}
+
+void tls_rec_set_ephemeral_key_size(TLS_REC *tls_rec, size_t size)
+{
+	g_assert(tls_rec != NULL);
+	tls_rec->ephemeral_key_size = size;
+}
+#endif
+
+void tls_rec_append_cert(TLS_REC *tls_rec, TLS_CERT_REC *tls_cert_rec)
+{
+	g_assert(tls_rec != NULL);
+	g_assert(tls_cert_rec != NULL);
+
+	tls_rec->certs = g_slist_append(tls_rec->certs, tls_cert_rec);
+}
+
+TLS_CERT_REC *tls_cert_create_rec()
+{
+	TLS_CERT_REC *rec = g_new0(TLS_CERT_REC, 1);
+	g_assert(rec != NULL);
+
+	return rec;
+}
+
+void tls_cert_rec_append_subject_entry(TLS_CERT_REC *tls_cert_rec, TLS_CERT_ENTRY_REC *tls_cert_entry_rec)
+{
+	g_assert(tls_cert_rec != NULL);
+	g_assert(tls_cert_entry_rec != NULL);
+
+	tls_cert_rec->subject = g_slist_append(tls_cert_rec->subject, tls_cert_entry_rec);
+}
+
+void tls_cert_rec_append_issuer_entry(TLS_CERT_REC *tls_cert_rec, TLS_CERT_ENTRY_REC *tls_cert_entry_rec)
+{
+	g_assert(tls_cert_rec != NULL);
+	g_assert(tls_cert_entry_rec != NULL);
+
+	tls_cert_rec->issuer = g_slist_append(tls_cert_rec->issuer, tls_cert_entry_rec);
+}
+
+void tls_cert_rec_free(TLS_CERT_REC *tls_cert_rec)
+{
+	if (tls_cert_rec == NULL)
+		return;
+
+	if (tls_cert_rec->subject != NULL) {
+		g_slist_free_full(tls_cert_rec->subject, (GDestroyNotify)tls_cert_entry_rec_free);
```

Comment:
`g_slist_free_full()` was introduced in glib 2.28. Current requirement is 2.16. Bumping it would mean dropping support for debian 6.0 (oldstable, squeeze, still supported). I suggest just iterating over the list.